File: examples/csharp/translateMath.cs
/**
 * @file    translateMath.cpp
 * @brief   Translates infix formulas into MathML and vice-versa
 * @author  Sarah Keating
 * @author  Ben Bornstein
 *
 *
 * This file is part of libSBML.  Please visit http://sbml.org for more
 * information about SBML, and the latest version of libSBML.
 */



using System;
using System.Collections.Generic;
using System.Text;
using libsbmlcs;

public class TranslateMath
{

    public static int Main(string[] args)
    {
        string line;
        string trimmed;
        string result;
        string str;
        long len;
        StringBuilder sb = new StringBuilder(1024);


        Console.WriteLine("This program translates infix formulas into MathML and");
        Console.WriteLine("vice-versa.  Enter or return on an empty line triggers");
        Console.WriteLine("translation. Ctrl-C quits");

        while (true)
        {
            Console.WriteLine("Enter infix formula or MathML expression (Ctrl-C to quit):");
            Console.Write("> ");

            line = Console.ReadLine();
            while (line != null)
            {
                trimmed = line.Trim();
                len = trimmed.Length;
                if (len > 0)
                {
                    sb.AppendLine(trimmed);
                }
                else
                {
                    str = sb.ToString();
                    result = (str[0] == '<') ? translateMathML(str) : translateInfix(str);

                    Console.WriteLine("Result:\n\n" + result + "\n\n");
                    sb = new StringBuilder(1024);
                    break;
                }

                line = Console.ReadLine();
            }
        }
        return 0;
    }


    /**
     * Translates the given infix formula into MathML.
     *
     * @return the MathML as a string.  The caller owns the memory and is
     * responsible for freeing it.
     */
    public static string translateInfix(string formula)
    {
        ASTNode math = libsbml.parseFormula(formula);

        string result = libsbml.writeMathMLToString(math);

        return result;
    }


    /**
     * Translates the given MathML into an infix formula.  The MathML must
     * contain no leading whitespace, but an XML header is optional.
     *
     * @return the infix formula as a string.  The caller owns the memory and
     * is responsible for freeing it.
     */
    public static string translateMathML(string xml)
    {
        ASTNode math = libsbml.readMathMLFromString(xml);
        string result = libsbml.formulaToString(math);
        return result;
    }
}

File: examples/r/translateMath.R
# 
# @file    translateMath.R
# @brief   Translates infix formulas into MathML and vice-versa
# @author  Frank Bergmann
# 
# 
# This file is part of libSBML.  Please visit http://sbml.org for more
# information about SBML, and the latest version of libSBML.
# 
# Usage: R --slave -f translateMath.R 
#
#

library(libSBML)

# Utility function to read a line from stdin
#
getline <- function() {
	f <- file("stdin")
	open(f)
	line = readLines(f,n=1)
	close(f)
	return (line)
}

# Utility function to trim the string
trim <- function(str) {
  return(gsub("(^ +)|( +$)", "", str))
}

# 
# Translates the given infix formula into MathML.
# 
# @return the MathML as a string.  The caller owns the memory and is
# responsible for freeing it.
# 
translateInfix <- function(formula) {

  math = parseFormula(formula);
  result = writeMathMLToString(math);
  
  return (result);
}


# 
# Translates the given MathML into an infix formula.  The MathML must
# contain no leading whitespace, but an XML header is optional.
# 
# @return the infix formula as a string.  The caller owns the memory and
# is responsible for freeing it.
# 
translateMathML <- function(xml) {
  # 
  # Prepend an XML header if not already present.
  #   
  if (substring(trim(xml),1, 2) != '<?') {
    header  = "<?xml version='1.0' encoding='UTF-8'?>\n";    
    math = readMathMLFromString(paste(header, xml));    
	return( formulaToString(math))
  } else {    
    math = readMathMLFromString(xml);
	return( formulaToString(math))
  }
}




cat( "\n" );
cat( "This program translates infix formulas into MathML and\n" );
cat( "vice-versa.  An 'enter' or a 'return' on an empty line\n" );
cat( "triggers translation. Ctrl-C quits\n" );
cat( "\n" );

while (TRUE) {
  cat( "Enter an infix formula or MathML expression (Ctrl-C to quit):\n" );
  cat( "\n" );
  cat( "> " );

  buffer = "" 
  
  repeat { 
    line = trim(getline());
    len  = nchar(line);

    if (len > 0) {
	  buffer = paste(buffer, line ,"\n", sep="");
    } else {
      if(substring(trim(buffer), 1,1) == "<") {
	    result = translateMathML(buffer)
      } else {
        result = translateInfix(buffer)
	  }
      cat("Result:\n\n",result,"\n\n\n");

      buffer = ""
	  break;
    }
  }
}

q(status=0);


